common/persistence/nosql/nosqlplugin/cassandra/workflowUtils.go

```diff
@@ -0,0 +1,916 @@
+			rowTypeExecutionTaskID)
+	}
+
+	// deleteInfos are the initiatedIDs for ChildInfo being deleted
+	for _, deleteInfo := range deleteInfos {
+		batch.Query(templateDeleteChildExecutionInfoQuery,
+			deleteInfo,
+			shardID,
+			rowTypeExecution,
+			domainID,
+			workflowID,
+			runID,
+			defaultVisibilityTimestamp,
+			rowTypeExecutionTaskID)
+	}
+	return nil
+}
+
+func (db *cdb) updateTimerInfos(
+	batch gocql.Batch,
+	shardID int,
+	domainID string,
+	workflowID string,
+	runID string,
+	timerInfos map[string]*persistence.TimerInfo,
+	deleteInfos []string,
+) error {
+	for _, timerInfo := range timerInfos {
+		batch.Query(templateUpdateTimerInfoQuery,
+			timerInfo.TimerID,
+			timerInfo.Version,
+			timerInfo.TimerID,
+			timerInfo.StartedID,
+			timerInfo.ExpiryTime,
+			timerInfo.TaskStatus,
+			shardID,
+			rowTypeExecution,
+			domainID,
+			workflowID,
+			runID,
+			defaultVisibilityTimestamp,
+			rowTypeExecutionTaskID)
+	}
+
+	for _, deleteInfo := range deleteInfos {
+		batch.Query(templateDeleteTimerInfoQuery,
+			deleteInfo,
+			shardID,
+			rowTypeExecution,
+			domainID,
+			workflowID,
+			runID,
+			defaultVisibilityTimestamp,
+			rowTypeExecutionTaskID)
+	}
+	return nil
+}
+
+func (db *cdb) updateActivityInfos(
+	batch gocql.Batch,
+	shardID int,
+	domainID string,
+	workflowID string,
+	runID string,
+	activityInfos map[int64]*persistence.InternalActivityInfo,
+	deleteInfos []int64,
+) error {
+	for _, a := range activityInfos {
+		batch.Query(templateUpdateActivityInfoQuery,
+			a.ScheduleID,
+			a.Version,
+			a.ScheduleID,
+			a.ScheduledEventBatchID,
+			a.ScheduledEvent.Data,
+			a.ScheduledTime,
+			a.StartedID,
+			a.StartedEvent.Data,
+			a.StartedTime,
+			a.ActivityID,
+			a.RequestID,
+			a.Details,
+			int32(a.ScheduleToStartTimeout.Seconds()),
+			int32(a.ScheduleToCloseTimeout.Seconds()),
+			int32(a.StartToCloseTimeout.Seconds()),
+			int32(a.HeartbeatTimeout.Seconds()),
+			a.CancelRequested,
+			a.CancelRequestID,
+			a.LastHeartBeatUpdatedTime,
+			a.TimerTaskStatus,
+			a.Attempt,
+			a.TaskList,
+			a.StartedIdentity,
+			a.HasRetryPolicy,
+			int32(a.InitialInterval.Seconds()),
+			a.BackoffCoefficient,
+			int32(a.MaximumInterval.Seconds()),
+			a.ExpirationTime,
+			a.MaximumAttempts,
+			a.NonRetriableErrors,
+			a.LastFailureReason,
+			a.LastWorkerIdentity,
+			a.LastFailureDetails,
+			a.ScheduledEvent.GetEncodingString(),
+			shardID,
+			rowTypeExecution,
+			domainID,
+			workflowID,
+			runID,
+			defaultVisibilityTimestamp,
+			rowTypeExecutionTaskID)
+	}
+
+	for _, deleteInfo := range deleteInfos {
+		batch.Query(templateDeleteActivityInfoQuery,
+			deleteInfo,
+			shardID,
+			rowTypeExecution,
+			domainID,
+			workflowID,
+			runID,
+			defaultVisibilityTimestamp,
+			rowTypeExecutionTaskID)
+	}
+	return nil
+}
+
+// NOTE: not sure we still need it. We keep the behavior for safe during refactoring
+// In theory we can just return the input as output
+// TODO: remove it in the future
```

Comment:
I think it's dropping everything smaller than a millisecond so that the time precision is one ms. I am not sure if any business logic is relying on it though.

Reply:
Yeah I never see any business rely on it. I feel like either this is some workaround of some Cassandra bugs, or the very first person wrote the code didn't pay much attention of what he's doing.

Reply:
I improved this comment to little bit so that next person won't get confused by my TODO.